<a href="https://colab.research.google.com/github/gregworks/python-for-data-science/blob/main/Intro-Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Did you install Anaconda?

- Install Anaconda (https://docs.anaconda.com/free/anaconda/install/)

> Make sure to install the latest version

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Python for Data Science

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/python.jpg?raw=1" width="300"  align="center"/>


<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# What is Python?
* a scripting language
* a programming language
* a command interpreter
* a dynamically typed language
* an object-oriented language

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Why Use Python?
* **P**opular
* eas**Y**
* fas**T** (sort of)
* pit**H**y
* br**O**ad
* efficie**N**t
* plus...
 * "batteries included"
 * _LARGE_ community
 * likable!

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

### Why Python?

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/whypython.png?raw=1" width="700" height="500" align="center"/>

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Why Not Use Python?
* CPU-bound applications can be faster in compiled languages (but not always!)
* real-time applications
* high-level–far from the metal

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## How to get around in Jupyter:
* Each place for you to enter text is called a _cell_
* Usually you enter __`Python`__ code, but you can also enter text in a _markup_ language called __`Markdown`__ (that's what's going on in _this_ cell)
* To "run" the code in the cell, hit __Shift-Return__ (i.e., hold down __Shift__ key, then hit __Return__)
* Try it with the cell below...

In [3]:
x = 6.34
x

4

* we'll work inside the Jupyter notebook and you'll be able to take it with you as a living, breathing document of your work in this class
* the __Insert__ menu will allow you to add a cell above or below the current cell
* the __Kernel__ menu will allow you to "talk" to the Python interpreter on your machine
  * (when you type into a cell, you are "talking" to the web browser, and the web browser sends the text to the __`Python`__ interpreter to be "run")
  * the __Kernel__ menu will allow you to _restart_ your __`Python`__ interpreter in case something goes wrong and it stops responding to you
  

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Variables/Typing
* no declarations
* basic data types are __int, float, string, boolean__
* everything is an object
* dynamically typed

In [6]:
x = 3
y = 24.99
print(y, x, "hello it's me")

24.99 3 hello it's me


In [7]:
print(x)
x = 'Prince'
x

3


'Prince'

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Strongly typed!

In [8]:
prince = 'Prince'

In [9]:
prince + 1999

TypeError: can only concatenate str (not "int") to str

In [10]:
prince + str(1999) # 'Prince' + '1999'

'Prince1999'

In [15]:
i = 1 # int
f = 1.4 # float for floating point number basically a decimal
b = True # bool (or boolean)
s = 'True' # string
b = 5 # int or integer
s + str(b)

'True5'

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Some [Builtin Python Functions](https://docs.python.org/3/library/functions.html)

## __`str()`__
* returns a string containing a nicely printable representation of the object passed as its argument

In [16]:
str(1999)

'1999'

In [17]:
str(True)

'True'

In [18]:
str(1.33e14)

'133000000000000.0'

In [19]:
str('x')

'x'

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## `int()`
* returns an integer object constructed from its argument–will be an error if not a number!

In [20]:
x = '503'
int(x)

503

In [21]:
x += 'a' # x = x + 'a' ... '503a'
int(x)

ValueError: invalid literal for int() with base 10: '503a'

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## __`type()`__
* returns the type of the object passed as its argument

In [23]:
x = 1
x, type(x)

(1, int)

In [24]:
x += 0.33 # x = x + 0.33
x, type(x)

(1.33, float)

In [25]:
type(True)

bool

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## __`print()`__
* what does this do?

In [26]:
name = 'Bruce Lee'
print(name)

Bruce Lee


In [27]:
a, b, c = 47, -12, 19
print(a, b, c)
print(a, b, c, sep=', ') # sep is "keyword argument"

47 -12 19
47, -12, 19


In [28]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Lab: str/int/type
*  use the Python interpreter/IDLE/Jupyter to type the following and think about the answer you expect before actually running the code
 * if result is not what you expect, that's an opportunity for learning

<pre><b>
str(53.3)
str(False)
str(false)
int('300')
int('30x')
type(False)
type('False')
type(3.5)
</b></pre>

In [1]:
int(false)

NameError: name 'false' is not defined

In [39]:
true

NameError: name 'true' is not defined

In [36]:
str(false)

NameError: name 'false' is not defined

In [29]:
print(str(53.3))
print(str(False))

53.3
False


In [30]:
type(str(False))

str

In [32]:
type(False)

bool

In [33]:
int('300')

300

In [34]:
int('30x')

ValueError: invalid literal for int() with base 10: '30x'

In [35]:
type(3.5)

float

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Python Arithmetic

* Python interpreter can perform arithmetic similar to other languages

In [2]:
3 / 2

1.5

In [3]:
3 // 2 # "int" division

1

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# div/mod/divmod

In [4]:
9 // 5 # "quotient"

1

In [5]:
9 % 5 # remainder when dividing 9 by 5

4

In [6]:
divmod(9, 5) # Python functions can and often do return multiple values

(1, 4)

In [7]:
quot, rem = divmod(9, 5)
print(quot)
print(rem)

1
4


<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Python has unlimited precision integers

In [8]:
2 * 3

6

In [9]:
2 ** 3

8

In [14]:
2 ** 20000

ValueError: Exceeds the limit (4300) for integer string conversion; use sys.set_int_max_str_digits() to increase the limit

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Floating Point Numbers
* similar to floating point numbers in other languages
* use __`float()`__ to convert to float, just as __`int()`__ converts to integer

In [15]:
x = 1
float(x)

1.0

In [16]:
float(True)

1.0

In [17]:
float('x')

ValueError: could not convert string to float: 'x'

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Strings

* use single or double quotes
* `\` lets you escape the next character, i.e., avoid its usual meaning

In [19]:
'hello it's me

SyntaxError: invalid syntax (<ipython-input-19-b04b4e0983d3>, line 1)

In [20]:
"hello it's me"

"hello it's me"

In [21]:
string1 = "This string isn't a problem"
string1

"This string isn't a problem"

In [22]:
string2 = 'This string is a "good" example'
string2

'This string is a "good" example'

In [27]:
string3 = 'This string isn\'t "more difficult" to read'
print(string3)

This string isn't "more difficult" to read


In [30]:
palindrome = 'A man,\tA plan,\nA canal:\nPanama.'
palindrome

'A man,\tA plan,\nA canal:\nPanama.'

In [31]:
print(palindrome)

A man,	A plan,
A canal:
Panama.


* `+` = concatenation operator
* `*` = duplication operator

In [32]:
s, t = "hello", 'bye' # bad practice
print(s + t)
print(s, t)

hellobye
hello bye


In [34]:
print(s * 4)
'-' * 75

hellohellohellohello


'---------------------------------------------------------------------------'

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Multi-Line Strings
* triple quotes allow for easy multi-line strings

In [37]:
''' a '''

' a '

In [38]:
s = """
isn't this a
multi-line string
?
"""

s # compute the value of this expression

"\nisn't this a\nmulti-line string\n?\n"

In [39]:
print(s)


isn't this a
multi-line string
?



<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Lab: Strings
* try these...
* ...be sure you understand what they do

```a, b, o, p = 'b', 'a', 'p', 'o'```

```o + p + o```

```a * 3 + b```

```a + p * 2 + 'k' * 2 + 'e' * 2 + o + 'er'```

In [40]:
a, b, o, p = 'b', 'a', 'p', 'o'



In [41]:
o + p + o




'pop'

In [42]:
a * 3 + b

'bbba'

In [43]:
a + p * 2 + 'k' * 2 + 'e' * 2 + o + 'er'



'bookkeeper'

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## __`len()`__
* returns the length of a string

In [44]:
p = 'Prince'
len(p)

6

In [45]:
len('')

0

In [48]:
len(p * 5)
print(p * 5)

PrincePrincePrincePrincePrince


<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Indexing Strings with __`[]`__
* access a single character via its offset
* easier to think of offset as opposed to index
* negative offsets count from end of string

In [49]:
alphabet = 'abcdefghijklmnopqrstuvwxyz'
alphabet[0]

'a'

In [51]:
alphabet[26]

IndexError: string index out of range

In [50]:
alphabet[25] # len(alphabet)-1

'z'

In [52]:
alphabet[len(alphabet)-1]

'z'

In [53]:
alphabet[-1] # idiomatic -- I want the last value or start from the end of the string and count backwards

'z'

In [58]:
len(alphabet)

26

In [60]:
alphabet[len(alphabet)-26]

'a'

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Let's start writing some code!

In [61]:
help(input)

Help on method raw_input in module ipykernel.kernelbase:

raw_input(prompt='') method of google.colab._kernel.Kernel instance
    Forward raw_input to frontends
    
    Raises
    ------
    StdinNotImplementedError if active frontend doesn't support stdin.



In [62]:
name = input('Enter your name: ')
print('You entered', name)

Enter your name: Greg
You entered Greg


<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Indentation

## Indentation
* colons and indentation delineate blocks {...}
* no braces!
* this will trip you up at first but once you're used to it, you'll love it

In [65]:
x = 2 # assign the value of 2 to x
if x == 1: # is x equal to 1?
    print('Hey, x is 1!')
    print('first part of if')
else:
    print(f'x is something other than 1, x is {x}')
    print('more stuff')

x is something other than 1, x is 2
more stuff


## Indentation (continued)
*  indentation must be consistent throughout the block

In [74]:
x = 1
if x == 1:
    print('x is 1')
    print('something else')

x is 1
something else


*  you can use any indentation you want as long as it's 4 spaces (PEP-8
https://www.python.org/dev/peps/pep-0008/)

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

<a id="functions-def"></a>
## Common Python Functions and Control Flow
---


<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/control.png?raw=1" width="800"  align="center"/>

---
In this section, we're going to tackle some common design patterns in Python. The first is the concept of control flow — this is how our programs will return different results based on specific input. Second, we'll cover basic functions — these let us create snippets of code that we can call later in a script, which creates code that's easier to read and maintain. Remember, we're going to be reading code much more often than writing it!

## `if… else` Statements

In Python, indentation matters! This is especially true when we look at the control structures in this lesson. In each case, a block of indented code is only run some of the time. There will always be a condition in the line preceding the indented block that determines whether the indented code is run or skipped.

### `if` Statement
The simplest example of a control structure is the `if` statement. We start with `if`, followed by something that can evaluate to `True` or `False` (such as any of the comparison operators we discussed earlier).

In [75]:
if 1 == 1:
    print('The integer 1 is equal to the integer 1.')
    print('Is the next indented line run, too?')

The integer 1 is equal to the integer 1.
Is the next indented line run, too?


In [78]:
if 'one' == 'two':
    print("The string 'one' is equal to the string 'two'.")

print('---')
print('These two lines are not indented, so they are always run next.')

---
These two lines are not indented, so they are always run next.


Notice that, in Python, the line before every indented block must end with a colon (`:`). In fact, it turns out that the `if` statement has a very specific syntax:

```
if <expression>:
    <one or more indented lines>
```

When the `if` statement is run, the expression is evaluated to `True` or `False` by applying the built-in `bool()` function. If the expression evaluates to `True`, the code block is run; otherwise, it is skipped.

#### Now You Try!
<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/hands_on.jpg?raw=1" width="100" height="100" align="right"/>


Create your own string called `test_string`, then fill in the blanks here to create an `if... else` statement for whether or not the first character in `test_string` is a lowercase `a`.

In [82]:
#### Type your answer here
test_string='Apple is a company'
if test_string[0] == "a":
  print("Yes")
else:
  print("No")


No


#### `if` ... `else`

In many cases, you may want to run some code if the expression evaluates to `True` and some other code if it evaluates to `False`. This is done using `else`. Note how it is at the same indentation level as the `if` statement, followed by a colon, followed by a code block. Let's see it in action.

In [84]:
if 50 < 30:
    print("50 < 30.")
else:
    print("50 >= 30.")
    print("The else code block was run instead of the first block.")

print('---')
print('These two lines are not indented, so they are always run next.')

50 >= 30.
The else code block was run instead of the first block.
---
These two lines are not indented, so they are always run next.


#### `if` ... `elif` ... `else`

Sometimes, you might want to run one specific code block out of several. For example, perhaps we provide the user with three choices and want something different to happen with each one.

`elif` stands for `else if`. It belongs on a line between the initial `if` statement and an (optional) `else`.

In [87]:
health = 18

if health > 70 :
    print('You are in great health!')
elif health > 40:
    print('Your health is average.')
    print('Exercise and eat healthily!')
elif health > 20:
    print('Your health is ok.')
    print('Dont eat meat!')
else:
    print('Your health is low.')
    print('Please see a doctor now.')

print('---')
print('These two lines are not indented, so they are always run next.')

Your health is low.
Please see a doctor now.
---
These two lines are not indented, so they are always run next.


This code works by evaluating each condition in order. If a condition evaluates to `True`, the rest are skipped.

**Let's walk through the code.** First, we let `health = 55`. We move to the next line at the same indentation level — the `if`. We evaluate `health > 70` to be `False`, so its code block is skipped. Next, the interpreter moves to the next line at the same outer indentation level, which happens to be the `elif`. It evaluates its expression, `health > 40`, to be `True`, so its code block is run. Now, because a code block was run, the rest of the `if` statement is skipped.

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Comparison Operators

| operator | meaning |
|---|---|
| == | equality  |
| != | inequality|
| < | less than |
| <= | less than or equals |
| > | greater than |
| >= | greater than or equals |
| in | membership |

In [88]:
x = 7

In [89]:
5 < x

True

In [90]:
x < 9

True

In [91]:
5 < x and x < 9 # &&

True

In [94]:
(5 < x) and (x < 9)

True

In [95]:
5 < x < 9

True

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Loops

* two kinds of loops in Python
 * __`while`__ loops ("do something until a condition becomes false")
 * __`for`__ loops ("do something a certain number of times")

# `while` loop example

In [98]:
import random # "batteries included"
# what do you think the line below does?
my_number = random.randint(1, 100)
guess = 0
# loop until...?

while guess != my_number:
    guess = int(input('Enter your guess (0 to give up): '))
    if guess == 0:
        print("Sorry that you're giving up!")
        break # abnormal termination
    elif guess > my_number:
        print("Guess was too high")
    elif guess < my_number:
        print("Guess was too low")
else:
    print("Congratulations. You guessed it!")
# break would put us here

Enter your guess (0 to give up): 99
Guess was too high
Enter your guess (0 to give up): 0
Sorry that you're giving up!


## `for` loop example
* typically used to cycle through an _iterable_ (string, list, and others we haven't learned yet) one element at a time
* "for thing in container"

In [103]:
a = [1,2,3,4,6, "aadsf"]
a
type(a)

for val in a:
  print(val)

1
2
3
4
6
aadsf


In [102]:
for letter in 'Python': # for each element in the container
    print(letter)

P
y
t
h
o
n


## `for` Loops
---
<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/for.png?raw=1" width="500" height="500" align="right"/>



One of the primary purposes of using a programming language is to automate repetitive tasks. One example is the `for` loop.

The `for` loop allows you to perform a task repeatedly on every element within an object, such as every name in a list.


Let's see how the pseudocode works:

```python
# For each individual object in the list
    # perform task_A on said object.
    # Once task_A has been completed, move to next object in the list.
```

Let's say we wanted to print each of the names in the list, as well as "is Awesome!" In this case, we'd create a temporary variable for each element in the collection (`for name in names` would put each name, in sequence, under the temporary variable `name`) and then do something with it.

In [104]:
names = ['Rebecca Bunch', 'Paula Proctor', 'Heather Davis']

for item in names:
    print(item + ' Is Awesome!')

Rebecca Bunch Is Awesome!
Paula Proctor Is Awesome!
Heather Davis Is Awesome!


We can also combine `if... else` statements and `for` loops:

In [105]:
names

['Rebecca Bunch', 'Paula Proctor', 'Heather Davis']

In [106]:
for item in names:
    if item == 'Paula Proctor':
        print(item + ' Is REALLY AWESOME!')
    else:
        print(item + ' Is Awesome!')

Rebecca Bunch Is Awesome!
Paula Proctor Is REALLY AWESOME!
Heather Davis Is Awesome!


#### Now You Try!
<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/hands_on.jpg?raw=1" width="100" height="100" align="right"/>
Create a new `if... elif... else` and `for` loop combination, using a list of your own choice.

In [110]:
117 % 2

1

In [111]:
### Type your answer here

nums=[1,23,4,6,7,76, 77,35]
# only print the even numbers

for num in nums:
  if num % 2 == 0:
    print(num)
  #else:
   # print('no odd numbers')


4
6
76


<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Sequences are also Iterable


In [115]:
list(range(1,11))

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [117]:
help(range)

Help on class range in module builtins:

class range(object)
 |  range(stop) -> range object
 |  range(start, stop[, step]) -> range object
 |  
 |  Return an object that produces a sequence of integers from start (inclusive)
 |  to stop (exclusive) by step.  range(i, j) produces i, i+1, i+2, ..., j-1.
 |  start defaults to 0, and stop is omitted!  range(4) produces 0, 1, 2, 3.
 |  These are exactly the valid indices for a list of 4 elements.
 |  When step is given, it specifies the increment (or decrement).
 |  
 |  Methods defined here:
 |  
 |  __bool__(self, /)
 |      True if self else False
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(self, key, /)
 |      Return self[key].
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __hash

In [116]:
for num in range(1, 10): # "for thing in container"
    print(num)

1
2
3
4
5
6
7
8
9


In [120]:
list(range(10, 0, -1))

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

In [125]:
for num in range(10, 0, -1):
    print(num, end='...')
print('blast off!')

10...9...8...7...6...5...4...3...2...1...blast off!


In [127]:
for num in range(-5, 6): # -5 ... 5
    if num == 0:
        continue # skip the remainder of the loop, and go to next iteration
    print(1 / num, end=' ')

-0.2 -0.25 -0.3333333333333333 -0.5 -1.0 1.0 0.5 0.3333333333333333 0.25 0.2 

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Quick Lab: Loops/Strings
* have the user enter a string, then loop through the string to generate (or print) a new string in which every character is duplicated, e.g., "Python" => "PPyytthhoonn"

In [128]:
name = input('Add a Company Name: ')

double_letter_company = ''

for letter in name:
  double_letter_company += letter *2

print(double_letter_company)

Add a Company Name: starbucks
ssttaarrbbuucckkss


In [129]:
entry = input("Please enter a string")
for c in entry:
  print(c * 2, end = "")

Please enter a stringstarbucks
ssttaarrbbuucckkss

## Lab: Loops
* Loop through the numbers from 2 to 25 and print out which numbers are prime, and for those numbers which are not prime numbers, you should print them as a product of two factors
* Remember that prime = no divisors other than 1 and itself
* Don't worry about efficiency, but if you're interested, check out math.sqrt()
* example output:
<pre>
2 is a prime number
3 is a prime number
4 equals 2 * 2
5 is a prime number
6 equals 2 * 3
7 is a prime number
8 equals 2 * 4
9 equals 3 * 3
10 equals 2 * 5
11 is a prime number
12 equals 2 * 6
13 is a prime number
14 equals 2 * 7
15 equals 3 * 5
16 equals 2 * 8
17 is a prime number
18 equals 2 * 9
19 is a prime number
20 equals 2 * 10
21 equals 3 * 7
22 equals 2 * 11
23 is a prime number
24 equals 2 * 12
25 equals 5 * 5
</pre>

In [132]:
for i in range(2,26):
    if i%2!=0:
      print(i, "is a prime number")
    elif i%2 == 0:
      print(i,"equals",i//2, "*", 2)

2 equals 1 * 2
3 is a prime number
4 equals 2 * 2
5 is a prime number
6 equals 3 * 2
7 is a prime number
8 equals 4 * 2
9 is a prime number
10 equals 5 * 2
11 is a prime number
12 equals 6 * 2
13 is a prime number
14 equals 7 * 2
15 is a prime number
16 equals 8 * 2
17 is a prime number
18 equals 9 * 2
19 is a prime number
20 equals 10 * 2
21 is a prime number
22 equals 11 * 2
23 is a prime number
24 equals 12 * 2
25 is a prime number


In [130]:
# start the lab here
for i in range(2, 26):
  factor = 0
  for f in range(2, i):
    if i % f == 0: #is it divisable?
      factor = f # is a factor
  if (factor == 0):
    print(f"{i} is a prime number")
  else:
    multiplier = i // factor
    print(f"{i} equals {multiplier} * {factor}")


2 is a prime number
3 is a prime number
4 equals 2 * 2
5 is a prime number
6 equals 2 * 3
7 is a prime number
8 equals 2 * 4
9 equals 3 * 3
10 equals 2 * 5
11 is a prime number
12 equals 2 * 6
13 is a prime number
14 equals 2 * 7
15 equals 3 * 5
16 equals 2 * 8
17 is a prime number
18 equals 2 * 9
19 is a prime number
20 equals 2 * 10
21 equals 3 * 7
22 equals 2 * 11
23 is a prime number
24 equals 2 * 12
25 equals 5 * 5


In [133]:
import math

def is_prime(n):
    """Check if a number is prime."""
    if n <= 1:
        return False
    if n <= 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True

def find_factors(n):
    """Find factors of a non-prime number."""
    factors = []
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            factors.append(i)
            factors.append(n // i)
            break
    return factors

# Loop through numbers from 2 to 25
for num in range(2, 26):
    if is_prime(num):
        print(f"{num} is a prime number")
    else:
        factors = find_factors(num)
        print(f"{num} equals {factors[0]} * {factors[1]}")

2 is a prime number
3 is a prime number
4 equals 2 * 2
5 is a prime number
6 equals 2 * 3
7 is a prime number
8 equals 2 * 4
9 equals 3 * 3
10 equals 2 * 5
11 is a prime number
12 equals 2 * 6
13 is a prime number
14 equals 2 * 7
15 equals 3 * 5
16 equals 2 * 8
17 is a prime number
18 equals 2 * 9
19 is a prime number
20 equals 2 * 10
21 equals 3 * 7
22 equals 2 * 11
23 is a prime number
24 equals 2 * 12
25 equals 5 * 5


<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Loops: Recap
* __`for`__ loop is more common
* __`break`__ exits loop immediately
* __`continue`__ skips remainder of loop and starts next iteration
* __`else`__ is executed if loop terminates normally (i.e., no __`break`__)

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Revisiting Strings

## Slices
* __`[start:stop:step]`__
* extracts the substring from __`start`__ to __`stop`__ _minus 1_, skipping __`step`__ characters at a time
* each of the st... are optional

In [134]:
alphabet = 'abcdefghijklmnopqrstuvwxyz'
         #  01234567890123456789012345
         #                         321-

In [135]:
alphabet[10:15]

'klmno'

In [136]:
alphabet[:5]

'abcde'

In [137]:
alphabet[23:]

'xyz'

In [138]:
alphabet[3:23:3]

'dgjmpsv'

In [139]:
alphabet[10:2:-1]

'kjihgfed'

In [140]:
alphabet[-3:]

'xyz'

In [141]:
alphabet[::-1] # idiomatic

'zyxwvutsrqponmlkjihgfedcba'

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## More String Functions (Methods)

In [142]:
poem = """TWO roads diverged in a yellow wood,
And sorry I could not travel both
And be one traveler, long I stood
And looked down one as far as I could
To where it bent in the undergrowth;

Then took the other, as just as fair,
And having perhaps the better claim,
Because it was grassy and wanted wear;
Though as for that the passing there
Had worn them really about the same,

And both that morning equally lay
In leaves no step had trodden black.
Oh, I kept the first for another day!
Yet knowing how way leads on to way,
I doubted if I should ever come back.

I shall be telling this with a sigh
Somewhere ages and ages hence:
Two roads diverged in a wood, and I—
I took the one less traveled by,
And that has made all the difference."""

In [143]:
len(poem) # built-in function

729

In [144]:
poem[:17]

'TWO roads diverge'

In [145]:
poem.startswith('TWO') # startswith is a function...a "method"
# NOT startswith(poem, 'TWO')

True

In [146]:
poem.endswith('And miles to go before I sleep.')

False

In [149]:
help(poem.endswith)

Help on built-in function endswith:

endswith(...) method of builtins.str instance
    S.endswith(suffix[, start[, end]]) -> bool
    
    Return True if S ends with the specified suffix, False otherwise.
    With optional start, test S beginning at that position.
    With optional end, stop comparing S at that position.
    suffix can also be a tuple of strings to try.



In [148]:
help(poem.find)

Help on built-in function find:

find(...) method of builtins.str instance
    S.find(sub[, start[, end]]) -> int
    
    Return the lowest index in S where substring sub is found,
    such that sub is contained within S[start:end].  Optional
    arguments start and end are interpreted as in slice notation.
    
    Return -1 on failure.



In [147]:
poem.find('the')

163

In [150]:
poem[163:178]

'the undergrowth'

In [152]:
help(poem.rfind)

Help on built-in function rfind:

rfind(...) method of builtins.str instance
    S.rfind(sub[, start[, end]]) -> int
    
    Return the highest index in S where substring sub is found,
    such that sub is contained within S[start:end].  Optional
    arguments start and end are interpreted as in slice notation.
    
    Return -1 on failure.



In [153]:
poem[714:]

'the difference.'

In [151]:
poem.rfind('the')

714

In [154]:
poem.count('the')

12

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## __`strip()`__

In [155]:
s = ' Now is the time      '
s.strip() # generates a new string in which leading/trailing...

'Now is the time'

In [156]:
s

' Now is the time      '

In [157]:
s = '.' + s.strip() + '...'

In [158]:
s

'.Now is the time...'

In [159]:
s.strip('.')

'Now is the time'

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Even More String Functions (Methods)...

In [161]:
help(s.capitalize)

Help on built-in function capitalize:

capitalize() method of builtins.str instance
    Return a capitalized version of the string.
    
    More specifically, make the first character have upper case and the rest lower
    case.



In [160]:
s = 'now IS the time'
s.capitalize()

'Now is the time'

In [163]:
help(s.title)

Help on built-in function title:

title() method of builtins.str instance
    Return a version of the string where each word is titlecased.
    
    More specifically, words start with uppercased characters and all remaining
    cased characters have lower case.



In [162]:
s.title()

'Now Is The Time'

In [164]:
s.upper()

'NOW IS THE TIME'

In [165]:
s.lower()

'now is the time'

In [167]:
help(s.swapcase)

Help on built-in function swapcase:

swapcase() method of builtins.str instance
    Convert uppercase characters to lowercase and lowercase characters to uppercase.



In [166]:
s.swapcase()

'NOW is THE TIME'

In [168]:
s

'now IS the time'

In [169]:
s.replace('the', 'not the') # be careful of the naming

'now IS not the time'

In [170]:
s.replace('t', 'T')

'now IS The Time'

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Lab: String Functions

* write a Python program which prompts the user for a string and a stride (increment), and alternately makes the string upper case and lower case, stride characters at a time, e.g.,
![alt-text](./images/uplow.png "uplow")


In [171]:
input_string = input("Enter a string: ")
stride = int(input("Enter the stride (increment): "))

result = []
upper = True  # Start with making characters uppercase

# Iterate through the input_string with the specified stride
for i in range(0, len(input_string), stride):
    if upper:
        result.append(input_string[i:i+stride].upper())
    else:
        result.append(input_string[i:i+stride].lower())

    # Toggle the case for the next segment
    upper = not upper

# Join the result segments into a modified string
modified_string = ''.join(result)

# Print the modified string
print("Modified string:", modified_string)

Enter a string: abcdefgh
Enter the stride (increment): 2
Modified string: ABcdEFgh


<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## __`split()/join()`__

In [172]:
'Now is the time'.split() # this is a string method

['Now', 'is', 'the', 'time']

In [173]:
'eggs, bread, milk, yogurt'.split(', ')

['eggs', 'bread', 'milk', 'yogurt']

In [174]:
['anti', 'dis', 'establish', 'men', 'tarian', 'ism']

['anti', 'dis', 'establish', 'men', 'tarian', 'ism']

In [178]:
# would be nice if we could write...
# ['a', 'b', 'c'].join(' ')
# but we don't because join is a string method
''.join(['anti', 'dis', 'establish', 'men', 'tarian', 'ism'])

'antidisestablishmentarianism'

In [179]:
', '.join(['Anne', 'Robert', 'Nancy'])

'Anne, Robert, Nancy'

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Lists

## Lists
* usually homogeneous, but may contain any objects
* unbounded / not a fixed size
* duplicates allowed
* __`list()`__ function creates a list from another sequence or container

In [180]:
mylist = [1, 3, 5, 7, 5, 3, 1]
mylist

[1, 3, 5, 7, 5, 3, 1]

In [181]:
days = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']
days

['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']

In [ ]:
list('hello')

In [182]:
date = '12/07/1941'
date.split('/')

['12', '07', '1941']

In [183]:
'a/b//c/d//e//f'.split('/')

['a', 'b', '', 'c', 'd', '', 'e', '', 'f']

In [184]:
stuff = input('Enter something: ')

Enter something: dogs cats


In [185]:
stuff.split()

['dogs', 'cats']

In [246]:
cars = ['Tesla', 'Fisker', 'Rivian', 'Lordstown']

In [188]:
cars[0]

'Tesla'

In [189]:
cars[-1] # always the last element of the list

'Lordstown'

In [190]:
vehicles = [cars, 'bus']
vehicles

[['Tesla', 'Fisker', 'Rivian', 'Lordstown'], 'bus']

In [191]:
vehicles[0]

['Tesla', 'Fisker', 'Rivian', 'Lordstown']

In [192]:
vehicles[0][-1]

'Lordstown'

In [194]:
testList = ['a']
testList

['a']

In [195]:
cars[-1] = 'Lordstown Motors'
cars

['Tesla', 'Fisker', 'Rivian', 'Lordstown Motors']

In [196]:
cars[:2] # first 2 items in a container

['Tesla', 'Fisker']

In [197]:
cars[::2] # the "evens", every other item

['Tesla', 'Rivian']

In [198]:
cars[1::2] # the "odds", every other item

['Fisker', 'Lordstown Motors']

In [199]:
cars[::-1] # also idiomatic

['Lordstown Motors', 'Rivian', 'Fisker', 'Tesla']

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Looping Through a List

In [203]:
index = 0
while index < len(cars):
    print(cars[index])
    index += 1 #i index = index + 1

Tesla
Fisker
Rivian
Lordstown Motors


In [204]:
for index in range(0, len(cars)):
    print(cars[index])

Tesla
Fisker
Rivian
Lordstown Motors


* that works, but it's not the way we'd write it in Python...it's not _Pythonic_

In [205]:
for car in cars: # for "thing in container"
    print(car)

Tesla
Fisker
Rivian
Lordstown Motors


In [206]:
for index, car in enumerate(cars):
  print(index, car)

0 Tesla
1 Fisker
2 Rivian
3 Lordstown Motors


In [209]:
for i, char in enumerate("abcd"):
  print(i, char)

0 a
1 b
2 c
3 d


<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Adding to a List ("mutator" methods)
* __`append()`__: add an item the end of a list
* __`insert()`__: add an item to a particular place in the list
* __`extend()`__ (also __`+=`__): add a list to a list

In [210]:
cars

['Tesla', 'Fisker', 'Rivian', 'Lordstown Motors']

In [211]:
cars.append('Lucid')
cars

['Tesla', 'Fisker', 'Rivian', 'Lordstown Motors', 'Lucid']

In [212]:
cars.insert(2, 'Faraday')
cars

['Tesla', 'Fisker', 'Faraday', 'Rivian', 'Lordstown Motors', 'Lucid']

In [213]:
others = ['Bollinger', 'Polestar']
cars += others # .extend()
cars

['Tesla',
 'Fisker',
 'Faraday',
 'Rivian',
 'Lordstown Motors',
 'Lucid',
 'Bollinger',
 'Polestar']

In [215]:
cars + others

['Tesla',
 'Fisker',
 'Faraday',
 'Rivian',
 'Lordstown Motors',
 'Lucid',
 'Bollinger',
 'Polestar',
 ['Bollinger', 'Polestar'],
 'Bollinger',
 'Polestar']

In [214]:
cars.append(others)
cars

['Tesla',
 'Fisker',
 'Faraday',
 'Rivian',
 'Lordstown Motors',
 'Lucid',
 'Bollinger',
 'Polestar',
 ['Bollinger', 'Polestar']]

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Removing from a List
* __`del`__: delete by position
* __`remove(item)`__: remove by value
* __`pop()`__: remove last item (or specified item)

In [216]:
cars

['Tesla',
 'Fisker',
 'Faraday',
 'Rivian',
 'Lordstown Motors',
 'Lucid',
 'Bollinger',
 'Polestar',
 ['Bollinger', 'Polestar']]

In [225]:
a = 1
print(a)
del a
print(a)

1


NameError: name 'a' is not defined

In [217]:
del cars[-1]
cars

['Tesla',
 'Fisker',
 'Faraday',
 'Rivian',
 'Lordstown Motors',
 'Lucid',
 'Bollinger',
 'Polestar']

In [218]:
cars.remove('Faraday')
cars

['Tesla',
 'Fisker',
 'Rivian',
 'Lordstown Motors',
 'Lucid',
 'Bollinger',
 'Polestar']

In [219]:
cars.pop() # last item by default

'Polestar'

In [220]:
cars

['Tesla', 'Fisker', 'Rivian', 'Lordstown Motors', 'Lucid', 'Bollinger']

In [221]:
cars.pop(1) # pop() or remove the second item

'Fisker'

In [223]:
cars.pop()

'Bollinger'

In [224]:
cars

['Tesla', 'Rivian', 'Lordstown Motors', 'Lucid']

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Examining Lists (inspectors)
* __`index(item)`__: return position of item
* __`count(item)`__: count occurrences of item
* __`in`__: test for membership

In [226]:
cars

['Tesla', 'Rivian', 'Lordstown Motors', 'Lucid']

In [227]:
cars.index('Lucid')

3

In [230]:
'Rivian' in cars

True

In [229]:
'Lordstown' in cars

False

In [231]:
for count in range(10): # do something 10 times
    cars.append('Byton')

In [232]:
cars

['Tesla',
 'Rivian',
 'Lordstown Motors',
 'Lucid',
 'Byton',
 'Byton',
 'Byton',
 'Byton',
 'Byton',
 'Byton',
 'Byton',
 'Byton',
 'Byton',
 'Byton']

In [233]:
cars.count('Byton')

10

In [234]:
while 'Byton' in cars:
    cars.remove('Byton') # each call only removes one
cars

['Tesla', 'Rivian', 'Lordstown Motors', 'Lucid']

In [235]:
for times in range(cars.count('Byton')):
    cars.remove('Byton')
cars

['Tesla', 'Rivian', 'Lordstown Motors', 'Lucid']

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## __`join()/split()`__–redux

In [236]:
cars

['Tesla', 'Rivian', 'Lordstown Motors', 'Lucid']

In [237]:
joined = ', '.join(cars)
joined # string which represents the "joined" items in the list

'Tesla, Rivian, Lordstown Motors, Lucid'

In [238]:
unjoined = joined.split(', ')
unjoined # split into a new list

['Tesla', 'Rivian', 'Lordstown Motors', 'Lucid']

In [239]:
cars == unjoined # are they the same? (They should be...)

True

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Sorting Lists
* __`sort()`__: _method_ to sort a list in place
* __`sorted()`__: _built-in function_ which returns a sorted list created
from an iterable/sequence
* __`len()`__: returns length of a list

In [241]:
help(id)

Help on built-in function id in module builtins:

id(obj, /)
    Return the identity of an object.
    
    This is guaranteed to be unique among simultaneously existing objects.
    (CPython uses the object's memory address.)



In [240]:
print(cars, id(cars))

['Tesla', 'Rivian', 'Lordstown Motors', 'Lucid'] 136661547825408


In [242]:
cars.sort() # mutator method which sorts the list
print(cars, id(cars))

['Lordstown Motors', 'Lucid', 'Rivian', 'Tesla'] 136661547825408


In [243]:
cars.sort(reverse=True)
print(cars, id(cars))

['Tesla', 'Rivian', 'Lucid', 'Lordstown Motors'] 136661547825408


In [244]:
# built-in function which can't mutate the list
sorted_list = sorted(cars)
print(sorted_list, id(sorted_list), id(cars))

['Lordstown Motors', 'Lucid', 'Rivian', 'Tesla'] 136661546320960 136661547825408


<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# List assignment does not copy the values!
* use __`copy()`__, __`list()`__, or __`[:]`__ to copy the values
* we'll use __`pythontutor.com`__ to understand the difference

In [247]:
cars

['Tesla', 'Fisker', 'Rivian', 'Lordstown']

In [248]:
c = cars
c[2] = 'Lordstown Motors'
cars

['Tesla', 'Fisker', 'Lordstown Motors', 'Lordstown']

In [249]:
c = cars.copy()
c[2] = 'Lordstown'
cars

['Tesla', 'Fisker', 'Lordstown Motors', 'Lordstown']

In [250]:
c

['Tesla', 'Fisker', 'Lordstown', 'Lordstown']

In [251]:
c = list(cars)
c

['Tesla', 'Fisker', 'Lordstown Motors', 'Lordstown']

In [252]:
c = cars[:]
c

['Tesla', 'Fisker', 'Lordstown Motors', 'Lordstown']

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Quick Lab: Lists
* Write a Python program to read in a list of items possibly containing duplicates, and then constructs a new list which contains the elements from the original list, with the order preserved, but the duplicates removed
![alt-text](images/list2.png "list2")

In [253]:
items = input("Enter items separated by spaces: ").split()

Enter items separated by spaces: cherry apple banana lemon apple apple cherry


In [254]:
list(set(items))

['apple', 'banana', 'lemon', 'cherry']

In [255]:
items = input("Enter items separated by spaces: ").split()

# Create an empty set to track unique elements
seen = set()
# Initialize an empty list to store unique elements in order
unique_items = []

# Iterate through the list of items
for item in items:
    # Check if the item has not been seen before (not in the set)
    if item not in seen:
        # Add the item to the set of seen elements
        seen.add(item)
        # Append the item to the unique_items list (to preserve order)
        unique_items.append(item)

# Print the resulting unique list
print("List with duplicates removed (order preserved):", unique_items)

Enter items separated by spaces: cherry apple banana lemon apple apple cherry
List with duplicates removed (order preserved): ['cherry', 'apple', 'banana', 'lemon']


In [ ]:
mylist = [1,2,3,3,4,5,6,6,8]

newlist = []
for item in mylist:
  print(item)
  if item not in newlist:
    newlist.append(item)
print(newlist)

## Lab: Lists
* Write a Python program to maintain a list
  * Read input until the user enters 'quit'
  * Words that the user enters should be added to the list
  * If a word begins with '-' (e.g., '-foo') it should be removed from the list
  * If the user enters only a '-', the list should be reversed
  * After each operation, print the list
  * Extras:
      * If user enters more than one word (e.g, __foo bar__), add "foo" and "bar" to the list, rather than "foo bar"
      * Same for "-", i.e., __-foo bar__ would remove "foo" and "bar" from the  list

In [ ]:
# do this on your own or at the end of class

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Dictionaries



# Dictionaries
* "unordered" grouping of key/value pairs
* sometimes called a "map", "hashmap", or "associative array"

In [256]:
d = {} # empty dict

In [257]:
d = { 'X': 10, 'V': 5, 'I': 1 } # can be initialized when declared

In [258]:
d

{'X': 10, 'V': 5, 'I': 1}

In [259]:
d['L'] = 50 # add something to the dict
print(d)

{'X': 10, 'V': 5, 'I': 1, 'L': 50}


In [260]:
# iterating through a dict iterates through the keys
for key in d: # for thing in container
    print(key, end=' ')

X V I L 

In [261]:
# ...of course we can print the values while iterating
for thing in d:
    print(thing, d[thing])

X 10
V 5
I 1
L 50


In [262]:
sbux_dict = {'venti': 20, 'tall': 12, 'grande': 16}
print(sbux_dict)

{'venti': 20, 'tall': 12, 'grande': 16}


In [263]:
print(sbux_dict.keys(), sbux_dict.values(),
      sbux_dict.items(), sep='\n')

dict_keys(['venti', 'tall', 'grande'])
dict_values([20, 12, 16])
dict_items([('venti', 20), ('tall', 12), ('grande', 16)])


In [264]:
total_ounces = 0
for amount in sbux_dict.values():
    total_ounces += amount

total_ounces

48

In [265]:
sum(sbux_dict.values())

48

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Dictionaries: View Objects
* __`keys()`__, __`values()`__, and __`items()`__ are view objects
* view objects provide a dynamic window into the dictionary

In [274]:
keys = sbux_dict.keys()
keys

dict_keys(['venti', 'tall', 'grande'])

In [275]:
# keys will change automagically after we add to the dict
print(keys)
sbux_dict['trenta'] = 31
print(keys)

dict_keys(['venti', 'tall', 'grande'])
dict_keys(['venti', 'tall', 'grande', 'trenta'])


In [276]:
keys

dict_keys(['venti', 'tall', 'grande', 'trenta'])

# __`get()`__: Dealing with missing dict values

In [277]:
d = {'foo': 'bar'}

In [278]:
d['foo']

'bar'

In [280]:
adfbadsf

NameError: name 'adfbadsf' is not defined

In [279]:
d['foot']

KeyError: 'foot'

In [281]:
if 'foot' in d: # is 'foot' a key in this dict
    print(d['foot'])
# or just... d.get('foot')

In [283]:
print(d.get('foo'))

bar


In [284]:
# what if we sort a dict?
for key in sorted(sbux_dict):
    print(key, sbux_dict[key])

grande 16
tall 12
trenta 31
venti 20


In [286]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [285]:
# In order to iterate in order, we have to sort the
# dict by value (as opposed to key)
# By default, sorted() will sort by key--
# usually not what we want!

for k in sorted(sbux_dict, key=sbux_dict.get):
    print(k, '=>', sbux_dict[k])

tall => 12
grande => 16
venti => 20
trenta => 31


<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Removing items from a dict
* __`del`__ = remove an item from the dict
* __`dict.pop(key)`__ = remove item and return value
* __`dict.clear()`__ = empty out the dict

In [287]:
mydict = {'trenta': 31, 'grande': 16, 'venti': 20,
          'tall': 12}
print(mydict)

{'trenta': 31, 'grande': 16, 'venti': 20, 'tall': 12}


In [288]:
del mydict['trenta']
print(mydict)

{'grande': 16, 'venti': 20, 'tall': 12}


In [289]:
print(mydict.pop('venti'))

20


In [290]:
print(mydict)

{'grande': 16, 'tall': 12}


In [291]:
mydict.clear()
mydict

{}

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Lab: dictionary
* use a dict to translate Roman numerals into their Arabic equivalents
1. load the dict with Roman numerals M (1000), D (500), C (100), L (50), X (10), V (5), I (1)
2. read in a Roman numeral
3. print Arabic equivalent
4. try it with MCLX = 1000 + 100 + 50 + 10 = 1160
4. __If you have time, deal with the case where a smaller number precedes a larger number, e.g., XC = 100 - 10 = 90, or MCM = 1000 + (1000-100) = 1900__
4. __MCMXCIX = 1999__

In [292]:
d = {'M': 1000, 'D': 500, 'C': 100, 'L': 50, 'X': 10, 'V': 5, 'I': 1}

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Dict Comprehension
* like a listcomp, a dictcomp creates a dict quickly

In [ ]:
names = ['Sally', 'Bob', 'Martha', 'Dirk']
employee_ids = [345, 286, 453, 119]
id_dict = { name: emp_id + 1000
                   for name, emp_id in zip(names, employee_ids)}
print(id_dict)

In [ ]:
d = { 'foo': 4, 'bar': -1, 'baz': -1, 'blah': 3, 'what': 2 }
print(d)

In [ ]:
d.items()

In [ ]:
d = { key: val for key, val in d.items()
               if val != -1 }
print(d)

In [ ]:
id_dict_inverse = { val : key for key, val in id_dict.items() }

In [ ]:
id_dict_inverse

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Now we understand this code!

In [ ]:
s = """Gur Mra bs Clguba, ol Gvz Crgref

Ornhgvshy vf orggre guna htyl.
Rkcyvpvg vf orggre guna vzcyvpvg.
Fvzcyr vf orggre guna pbzcyrk.
Pbzcyrk vf orggre guna pbzcyvpngrq.
Syng vf orggre guna arfgrq.
Fcnefr vf orggre guna qrafr."""

d = {}
for c in (65, 97):
    for i in range(26):
        d[chr(i+c)] = chr((i+13) % 26 + c)

print("".join([d.get(c, c) for c in s]))

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Functions

## Functions
* __`def`__ introduces a function, followed by function name, parenthesized list of args and then a colon
* body of function is indented

In [ ]:
# a "do nothing" function
def noop():
    pass # Python statement that does nothing

In [ ]:
noop()

In [ ]:
noop(1, 3, 5)

In [ ]:
def simpfunc(x):
    if x == 1:
        print('Hey, x is 1')
    elif x < 10:
        print('x is < 10 and not 1')
    else:
        print('x >= 10')

In [ ]:
simpfunc(1)

In [ ]:
simpfunc(5)

In [ ]:
simpfunc(15)

In [ ]:
simpfunc(2.4)

In [ ]:
def rounder25(amount):
    """Return amount rounded UP to nearest
       quarter dollar.

           ...$1.89 becomes $2.00
           ...but $1.00/$1.25/$1.75/etc.
           remain unchanged
    """
    dollars = int(amount) # 1
    cents = round((amount - dollars) * 100) # 89
    quarters = cents // 25 # 3
    if cents % 25: # 14
        quarters += 1 # 4
    amount = dollars + 0.25 * quarters # 2.00

    return amount

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Functions (cont'd)
* __`help(func)`__ prints out formatted docstring
* __`func`__ .\__doc__ prints out raw docstring

In [ ]:
help(rounder25)

In [ ]:
print(rounder25.__doc__)

In [ ]:
rounder25(1.89)

In [ ]:
rounder25(1.75)

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Functions (cont'd)
* if a function doesn’t call return explicitly, the special value __`None`__ is returned
* __`None`__ is like __`NULL`__ in other languages
* acts like __`False`__...but not the same as __`False`__

In [ ]:
retval = noop()
print(retval)

In [ ]:
# None acts like False...
if retval:
    print('something')
else:
    print('nothing')

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Functions: positional arguments
* arguments are passed to functions in order written
* downside: you must remember meaning of each position

In [ ]:
def menu(wine, entree, dessert):
    return { 'wine': wine, 'entree': entree,
            'dessert': dessert }

![alt-text](images/IDE.png "IDE")
* outside an IDE, it can be difficult to remember
* if you pass args in wrong order, bad things can happen!

In [ ]:
menu('chianti', 'tartuffo', 'polenta')

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Functions: keyword arguments
* you may specify arguments by name, in any order
* once you specify a keyword argument, all arguments following it must be keyword arguments

In [ ]:
# passing some arguments by keyword
menu('chianti', dessert='tartufo', entree='polenta')

In [ ]:
# passing all arguments by keyword
menu(dessert='tartufo', entree='polenta', wine='chianti')

In [ ]:
# once you start passing arguments by keyword, the rest must be passed by keyword
menu('chianti', dessert='tartufo', 'polenta')

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Functions: default arguments

In [ ]:
def menu(wine, entree, dessert='tartufo'):
    return { 'wine': wine, 'entree': entree,
            'dessert': dessert }

In [ ]:
menu('chardonnay', 'braised tofu')

In [ ]:
menu('chardonnay', dessert='canoli',
     entree='fagioli')

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Lab: functions
* Write a function __`calculate`__ which is passed two operands and an operator and returns the calculated result, e.g., __`calculate(2, 4, '+')`__ would return 6
* Write a function that, given a string, returns True or False whether the string is a pangram
* Write a function which takes an integer as a parameter, and sums up its digits. If the resulting sum contains more than 1 digit, the function should sum the digits again, e.g., __`sumdigits(1235)`__ should compute the sum of 1, 2, 3, and 5 (11), then compute the sum of 1 and 1, returning 2.
* Write a function which takes a number as a parameter and returns a string version of the number with commas representing thousands, e.g., __`add_commas(12345)`__ would return "12,345"
* Write a function to demonstrate the Collatz Conjecture:
  * for integer n > 1
    * if n is even, then __`n = n // 2`__
    * if n is odd, then __`n = n * 3 + 1`__
  * ...will always converge to 1
  * (your function should take n and keep printing new value of n until n is 1)


<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Variable Positional Arguments
* sometimes we want a function which takes a variable number of arguments (e.g., builtin __`print()`__ function)

In [ ]:
def func(*args): # func takes 0 or more arguments
    print(args)
    for index, arg in enumerate(args):
        print('arg', index, 'is', arg)

In [ ]:
func()

In [ ]:
func(3, 4, 5, [2, 2, 3], {}, 'string')

In [ ]:
func({ 'a': 'b'}, [1, 2, 3], 'this', True)

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Lab: Variable Positional Arguments
* write a function called __`product`__ which accepts a variable number of arguments and returns the product of all of its args. With no args, __`product()`__ should return 1    

<pre><b>
>>> product(3, 5)
15
>>> product(1, 2, 3)
6
>>> product(63, 12, 3, 0, 9)
0
>>> product()
1
</b></pre>

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Variable Keyword Arguments
* what if a function needs a bunch of configuration options, having default values which typically aren't overridden?
* one way to do this would be to have the function accept a dict in which these value(s) can be specified
* better way is to use variable keywords arguments

In [ ]:
def vka(**kwargs):
    print(kwargs)
    for key in kwargs:
        print(key, '=>', kwargs[key])

In [ ]:
vka(sep='+', foo='bar', whizbang='rotunda', x=5, debug='hello', color='pink')

In [ ]:
def weird_func(x, y, z, *args, **kwargs):
    print('req args:', x, y, z)
    print('var pos args', args)
    print('var keywd args', kwargs)
    if 'debug' in kwargs:
        if kwargs['debug'] == True: # because it could be false
            turn_on_debugging = True
            # utilize some of *args...

In [ ]:
def weird_func(x, y, z, debug_file=None, debug=False):
    print('req args:', x, y, z)
    print('var pos args', args)
    print('var keywd args', kwargs)

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

# Lab: Variable Keyword Arguments
* modify your __`calculate`__ function by adding variable keywords arguments to it and checking whether __`float = True`__, and if so, the calculation is done as floating point, rather than integer (of course this could be done with a default argument value, but don't do that)

<pre><b>
calculate(2, 4, '+') = 6
calculate(3, 2, '/', float=True) = 1.5
</b></pre>

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

## Functions: recap
* Python encourages functions which support lots of arguments with default values
* "Explicit is better than implicit"
  * arguments can be passed out of order ONLY if they're passed by keyword
  * keywords are more explicit than positions because the function call documents the purpose of its arguments
* variable positional args (__`*args`__)
* variable keyword args (__`**kwargs`__)

<img src="https://github.com/gregworks/python-for-data-science/blob/main/images/border.jpg?raw=1" height="10" width="1500" align="center"/>

**Exercise: Student Grade Calculator**

**Objective:** Create a Python program that calculates the final grades of students based on their exam scores, assignments, and participation.

**Instructions:**

1. Ask the user for the number of students for whom they want to calculate grades.
2. For each student, do the following:
   - Prompt the user for the student's name.
   - Prompt the user for the student's exam score (out of 100).
   - Prompt the user for the total assignment score (out of 100).
   - Prompt the user for the participation score (out of 100).
3. Calculate the final grade for each student using the following formula:
   - Final Grade = (Exam Score * 0.4) + (Assignment Score * 0.3) + (Participation Score * 0.3)
4. Print out the final grades for all students.

**Tips:**

- Use a loop to iterate through each student's data.
- Create variables to store the student's name, exam score, assignment score, and participation score.
- Define a function to calculate the final grade.
- Use `input()` to get user input and convert the input to the appropriate data types.
- Use control flow (if statements) to ensure that user inputs are within the valid range (0-100).
- Display the final grades for each student after all data is collected.

This exercise will help your students practice creating variables, using operators and functions, and implementing control flow in Python while working with data. It's a simple example of data manipulation that can be a foundation for more complex data science tasks later on.


<!-- def calculate_final_grade(exam_score, assignment_score, participation_score):
    # Calculate the final grade using the given formula
    final_grade = (exam_score * 0.4) + (assignment_score * 0.3) + (participation_score * 0.3)
    return final_grade

# Prompt the user for the number of students
num_students = int(input("Enter the number of students: "))

# Initialize a list to store student data
student_data = []

# Loop through each student
for i in range(num_students):
    print(f"\nStudent {i + 1}:")
    
    # Prompt for student name
    student_name = input("Enter student name: ")
    
    # Prompt for exam score, assignment score, and participation score
    exam_score = float(input("Enter exam score (out of 100): "))
    assignment_score = float(input("Enter assignment score (out of 100): "))
    participation_score = float(input("Enter participation score (out of 100): "))
    
    # Ensure scores are within the valid range (0-100)
    if 0 <= exam_score <= 100 and 0 <= assignment_score <= 100 and 0 <= participation_score <= 100:
        final_grade = calculate_final_grade(exam_score, assignment_score, participation_score)
        
        # Store student data
        student_data.append((student_name, final_grade))
    else:
        print("Invalid input. Scores must be between 0 and 100.")

# Display the final grades for all students
print("\nFinal Grades:")
for student_name, final_grade in student_data:
    print(f"{student_name}: {final_grade:.2f}") -->
